File: backend/app/api/routes/directory.py
"""Public directory endpoints — the client view (search & profiles)."""
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Query

from app.api.dependencies import DirectoryServiceDep
from app.schemas.common import Page
from app.schemas.firm import FirmRead, FirmSummary
from app.schemas.lawyer import LawyerRead, LawyerSummary
from app.schemas.practice_area import PracticeAreaRead

router = APIRouter(tags=["directory"])

PageParam = Annotated[int, Query(ge=1)]
PageSizeParam = Annotated[int, Query(ge=1, le=100)]


@router.get("/practice-areas", response_model=list[PracticeAreaRead])
async def list_practice_areas(service: DirectoryServiceDep) -> list[PracticeAreaRead]:
    return await service.list_practice_areas()


@router.get("/lawyers", response_model=Page[LawyerSummary])
async def search_lawyers(
    service: DirectoryServiceDep,
    practice_area_id: Annotated[int | None, Query()] = None,
    q: Annotated[str | None, Query(description="Busca textual")] = None,
    page: PageParam = 1,
    page_size: PageSizeParam = 20,
) -> Page[LawyerSummary]:
    return await service.search_lawyers(
        practice_area_id=practice_area_id, query=q, page=page, page_size=page_size
    )


@router.get("/lawyers/{lawyer_id}", response_model=LawyerRead)
async def get_lawyer(lawyer_id: int, service: DirectoryServiceDep) -> LawyerRead:
    return await service.get_lawyer(lawyer_id)


@router.get("/firms", response_model=Page[FirmSummary])
async def search_firms(
    service: DirectoryServiceDep,
    practice_area_id: Annotated[int | None, Query()] = None,
    q: Annotated[str | None, Query(description="Busca textual")] = None,
    page: PageParam = 1,
    page_size: PageSizeParam = 20,
) -> Page[FirmSummary]:
    return await service.search_firms(
        practice_area_id=practice_area_id, query=q, page=page, page_size=page_size
    )


@router.get("/firms/{firm_id}", response_model=FirmRead)
async def get_firm(firm_id: int, service: DirectoryServiceDep) -> FirmRead:
    return await service.get_firm(firm_id)

File: frontend/src/api/directory.ts
import { api } from "./client";
import type {
  FirmDetail,
  FirmSummary,
  LawyerDetail,
  LawyerSummary,
  Page,
  PracticeArea,
} from "../types";

export interface SearchParams {
  practice_area_id?: number | null;
  q?: string | null;
  page?: number;
  page_size?: number;
}

function cleanParams(params: SearchParams): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (params.practice_area_id) out.practice_area_id = params.practice_area_id;
  if (params.q) out.q = params.q;
  if (params.page) out.page = params.page;
  if (params.page_size) out.page_size = params.page_size;
  return out;
}

export async function listPracticeAreas(): Promise<PracticeArea[]> {
  const { data } = await api.get<PracticeArea[]>("/practice-areas");
  return data;
}

export async function searchLawyers(params: SearchParams): Promise<Page<LawyerSummary>> {
  const { data } = await api.get<Page<LawyerSummary>>("/lawyers", { params: cleanParams(params) });
  return data;
}

export async function getLawyer(id: number): Promise<LawyerDetail> {
  const { data } = await api.get<LawyerDetail>(`/lawyers/${id}`);
  return data;
}

export async function searchFirms(params: SearchParams): Promise<Page<FirmSummary>> {
  const { data } = await api.get<Page<FirmSummary>>("/firms", { params: cleanParams(params) });
  return data;
}

export async function getFirm(id: number): Promise<FirmDetail> {
  const { data } = await api.get<FirmDetail>(`/firms/${id}`);
  return data;
}
